pyhf/__init__.py

```diff
@@ -171,79 +126,57 @@ def set_poi(self,name):
         assert s.stop-s.start == 1
         self.poi_index = s.start
 
-    def add_modifier(self, name, npars, modifier, suggested_init, suggested_bounds):
-        is_constraint = type(modifier) in [histosys_constraint, normsys_constraint, shapesys_constraint]
-        if name in self.par_map:
-            if type(modifier) == normsys_constraint:
-                log.info('accepting existing normsys')
-                return False
-            if type(modifier) == histosys_constraint:
-                log.info('accepting existing histosys')
-                return False
-            if type(modifier) == type(None):
-                log.info('accepting existing unconstrained factor ')
-                return False
-            raise RuntimeError(
-                'shared systematic not implemented yet (processing {})'.format(name))
-        log.info('adding modifier %s (%s new nuisance parameters)', name, npars)
-
+    def add_or_get_modifier(self, channel, sample, modifier_def):
+        """
+        Add a new modifier if it does not exist and return it
+        or get the existing modifier and return it
+
+        Args:
+            channel: current channel object (e.g. from spec)
+            sample: current sample object (e.g. from spec)
+            modifier_def: current modifier definitions (e.g. from spec)
+
+        Returns:
+            modifier object
+
+        Example:
+            modifier = instance.add_or_get_modifier(channel, sample, modifier_def)
+        """
+        # check if modifier type is implemented
+        if modifier_def['type'] not in modifiers.registry:
+            raise RuntimeError('modifier type not implemented yet (processing {0:s})'.format(modifier_def['type']))
+
+        # get modifier class associated with modifier type
+        modifier_cls = modifiers.registry.get(modifier_def['type'])
+
+        # check if modifier of the given name already exists
+        if modifier_def['name'] in self.par_map:
+            if modifier_cls.is_constrained:
+                log.info('accepting existing {0:s} (type: {1:s})'.format(modifier_def['name'], modifier_cls.__name__))
+            else:
+                log.info('accepting existing unconstrained factor')
+            return self.par_map[modifier_def['name']]['modifier']
+
```

Comment:
This seems to *always* accept any shared modifier (e.g. if the first `if` clause passes). Before we had a filter to only accept shared modifiers if they are eiter a 

* histosys constrained modifier
* normsys constrained modifier
* any type of unconstrained modifier (e.g. shapefactor, normfactor)

notably shapesys should *not* be shared. Perhaps we should add this as a test

Reply:
Yes, I noticed this. I refactored it this way, because the refactoring is going to need to take a few steps. I want to switch to using singletons which might be more elegant here (http://python-3-patterns-idioms-test.readthedocs.io/en/latest/Singleton.html) as this removes the actual logic of knowing which modifier is shared and which is not, from `hfpdf` as it should not care -- this is modifier-specific behavior, no? I think something like the `Borg` version is the most elegant via something like

```
class Modifier(object):
  pass

class UnsharedModifier(Modifier):
  def __init__(self, name):
      pass

class SharedModifier(Modifier):
  _shared_state = {}
  def __init__(self, name):
    self.__dict__ = _shared_state.setdefault(name, {}).get(name)

class ShapeSys(UnsharedModifier):
  def __init__(self, name):
    super(self.__class__, self).__init__(name)
    self.name = name

class HistoSys(SharedModifier):
  def __init__(self, name):
    super(self.__class__, self).__init__(name)
    self.name = name
```

then the modifiers, themselves, encode whether or not they're shared -- and the `hfpdf` code doesn't care at all [and it shouldn't].

Reply:
Thinking again, the refactor should preserve existing functionality. I will file a new issue about refactoring later for singletons (if so desired). And I will fix this for now.

Reply:
yes I think we should always keep master functional. thanks!

Reply:
done